File: go/src/args.go
package main

import flag "github.com/spf13/pflag"
import "runtime"
import "fmt"
import "os"

type Args struct {
	DebugCpu bool
	DebugApu bool
	DebugGpu bool
	DebugRam bool
	Headless bool
	Silent   bool
	Turbo    bool
	Frames   int
	Profile  int
	Rom      string
}

func NewArgs() *Args {
	var debug_cpu = flag.BoolP("debug-cpu", "c", false, "Debug CPU")
	var debug_gpu = flag.BoolP("debug-gpu", "g", false, "Debug GPU")
	var debug_apu = flag.BoolP("debug-apu", "a", false, "Debug APU")
	var debug_ram = flag.BoolP("debug-ram", "r", false, "Debug RAM")
	var headless = flag.BoolP("headless", "H", false, "No video")
	var silent = flag.BoolP("silent", "S", false, "No audio")
	var turbo = flag.BoolP("turbo", "t", false, "No sleep()")
	var frames = flag.IntP("frames", "f", 0, "Exit after N frames")
	var profile = flag.IntP("profile", "p", 0, "Exit after N seconds")
	var version = flag.BoolP("version", "v", false, "Show build info")
	flag.Parse()
	if *version {
		fmt.Println(runtime.Version())
		os.Exit(0)
	}
	var rom = flag.Arg(0)

	return &Args{
		DebugCpu: *debug_cpu,
		DebugGpu: *debug_gpu,
		DebugApu: *debug_apu,
		DebugRam: *debug_ram,
		Headless: *headless,
		Silent:   *silent,
		Turbo:    *turbo,
		Frames:   *frames,
		Profile:  *profile,
		Rom:      rom,
	}
}

File: zig/src/args.zig
const clap = @import("clap");
const std = @import("std");
const builtin = @import("builtin");

const errors = @import("errors.zig");

const io = std.io;

pub const Args = struct {
    rom: []const u8,
    headless: bool,
    silent: bool,
    debug_cpu: bool,
    debug_gpu: bool,
    debug_apu: bool,
    debug_ram: bool,
    frames: u32,
    profile: u32,
    turbo: bool,

    pub fn parse_args() !Args {
        const params = comptime [_]clap.Param(clap.Help){
            clap.parseParam("-h, --help             Display this help and exit.") catch unreachable,
            clap.parseParam("-H, --headless         Disable GUI") catch unreachable,
            clap.parseParam("-S, --silent           Disable Sound") catch unreachable,
            clap.parseParam("-c, --debug-cpu        Debug CPU") catch unreachable,
            clap.parseParam("-g, --debug-gpu        Debug GPU") catch unreachable,
            clap.parseParam("-a, --debug-apu        Debug APU") catch unreachable,
            clap.parseParam("-r, --debug-ram        Debug RAM") catch unreachable,
            clap.parseParam("-f, --frames <u32>     Exit after N frames") catch unreachable,
            clap.parseParam("-p, --profile <u32>    Exit after N seconds") catch unreachable,
            clap.parseParam("-t, --turbo            No sleep()") catch unreachable,
            clap.parseParam("-v, --version          Show build info") catch unreachable,
            clap.parseParam("<str>                  ROM filename") catch unreachable,
        };

        var diag = clap.Diagnostic{};
        var res = clap.parse(clap.Help, &params, clap.parsers.default, .{
            .diagnostic = &diag,
        }) catch |err| {
            // Report useful error and exit
            diag.report(io.getStdErr().writer(), err) catch {};
            return err;
        };
        defer res.deinit();

        if (res.args.version != 0) {
            try std.io.getStdOut().writer().print("{s} {s} {s}\n", .{
                builtin.zig_version_string,
                @tagName(builtin.mode),
                @tagName(builtin.zig_backend),
            });
            return errors.ControlledExit.Help;
        }
        if (res.args.help != 0) {
            try clap.help(std.io.getStdErr().writer(), clap.Help, &params, .{});
            return errors.ControlledExit.Help;
        }

        var frames: u32 = 0;
        if (res.args.frames) |n|
            frames = n;

        var profile: u32 = 0;
        if (res.args.profile) |n|
            profile = n;

        var rom: []const u8 = "";
        for (res.positionals) |pos|
            rom = pos;

        return Args{
            .rom = rom,
            .headless = res.args.headless != 0,
            .silent = res.args.silent != 0,
            .debug_cpu = res.args.@"debug-cpu" != 0,
            .debug_gpu = res.args.@"debug-gpu" != 0,
            .debug_apu = res.args.@"debug-apu" != 0,
            .debug_ram = res.args.@"debug-ram" != 0,
            .frames = frames,
            .profile = profile,
            .turbo = res.args.turbo != 0,
        };
    }
};
